src/Sparrow/Json/BlittableWriter.cs

```diff
@@ -369,100 +403,339 @@ public void WriteNumber(int value, int sizeOfValue)
         {
-            // see zig zap trick here:
-            // https://developers.google.com/protocol-buffers/docs/encoding?csw=1#types
-            // for negative values
-
-            var buffer = _innerBuffer.Address;
-            var count = 0;
+            // We will do zigzag encoding for int64, but not for int32
             var v = (ulong)((value << 1) ^ (value >> 63));
-            while (v >= 0x80)
+
+            // If value is 0, write a single byte and return
+            if (v < 0x80)
             {
-                buffer[count++] = (byte)(v | 0x80);
-                v >>= 7;
+                _unmanagedWriteBuffer.Write((byte)v);
+                return 1;
             }
-            buffer[count++] = (byte)(v);
 
-            if (count == 1)
-                _unmanagedWriteBuffer.WriteByte(*buffer);
-            else
-                _unmanagedWriteBuffer.Write(buffer, count);
-
-            return count;
+            byte* dest = stackalloc byte[VariableSizeEncoding.MaximumSizeOf<long>()];
+            var writtenBytes = VariableSizeEncoding.Write(dest, v);
+            _unmanagedWriteBuffer.Write(dest, writtenBytes);
+            return writtenBytes;
         }
 
+#if NET6_0_OR_GREATER
+        [SkipLocalsInit]
+#endif
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public unsafe int WriteVariableSizeInt(int value)
         {
-            // assume that we don't use negative values very often
-            var buffer = _innerBuffer.Address;
+            // If value is 0, write a single byte and return
+            if (value is >= 0 and < 0x80)
+            {
+                _unmanagedWriteBuffer.Write((byte)value);
+                return 1;
+            }
+
+            byte* dest = stackalloc byte[VariableSizeEncoding.MaximumSizeOf<int>()];
+            var writtenBytes = VariableSizeEncoding.Write(dest, value);
+            _unmanagedWriteBuffer.Write(dest, writtenBytes);
+            return writtenBytes;
+        }
+
+#if NET6_0_OR_GREATER
+        [SkipLocalsInit]
+#endif
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe int WriteVariableSizeIntInReverse(int value)
+        {
+            // If value is 0, write a single byte and return
+            if (value is >= 0 and < 0x80)
+            {
+                _unmanagedWriteBuffer.Write((byte)value);
+                return 1;
+            }
+
+            // Calculate the number of bytes needed
+            int significantBits = 32 - Bits.LeadingZeroes((uint)value);
+            int byteCount = (significantBits + 6) / 7;
+
+            byte* dest = stackalloc byte[VariableSizeEncoding.MaximumSizeOf<int>()];
+            
+            byte* destPtr = dest + byteCount - 1;
 
-            var count = 0;
             var v = (uint)value;
             while (v >= 0x80)
             {
-                buffer[count++] = (byte)(v | 0x80);
+                *destPtr = (byte)(v | 0x80);
+                destPtr--;
```

Comment:
You're doing `destPtr--` now, while previous impl of `int WriteVariableSizeIntInReverse(int value)` did `buffer[count++] = ...`. In each iteration the same operations are performed:
- `(byte)(v | 0x80)`
- `v >>= 7;`

Didn't it change the format effectively? I don't see any failures but my concern is that it's potentially breaking change (if my understanding is valid)

Reply:
The algorithm is different now. We actually check beforehand how many bytes we are going to use and write it backwards.

Reply:
Ok, let me rephrase - does it change the blittable format (raw data on disk)?

Reply:
No. We just process it differently.

Reply:
OK